MAJOR: matching parameters in InMemoryQueueManager to parent class in event-dispatcher

spekulatius opened the pull request:
Fixes #66

solverat commented:
This is a huge bc break and should get merged in a new major release only. However, i think there are some more dispatching events to change.

mvdbos commented:
First of all, thanks for contributing. I had actually forgotten that I set myself the task to fix this in November, as can be seen in #61, and in #62. 

For some reason the build feedback from Travis is not working, but it is failing. Please check here: https://travis-ci.org/github/mvdbos/php-spider/jobs/666992620#L280

This is because there are more places to fix this event dispatcher invocation. Can you run the tests locally and make sure everything passes?

mvdbos commented:
> This is a huge bc break and should get merged in a new major release only. 

Agreed. That said, I am hesitant to go to version 1. I do not pretend that this is production quality code in any way. ;-) But I fear there is no alternative if we want users with a version constraint like `^0.2`to be protected against BC breaks .

spekulatius commented:
Hello @mvdbos and @solverat,

Thanks for replying to the PR. I agree - it's a major change. From a user' view the interface doesn't change, does it?

I'll check the tests locally and see how far I get with addressing the issues. Once I'm through I'll report back.

Cheers,
Peter

spekulatius commented:
> > This is a huge bc break and should get merged in a new major release only.
> 
> Agreed. That said, I am hesitant to go to version 1. I do not pretend that this is production quality code in any way. ;-) But I fear there is no alternative if we want users with a version constraint like `^0.2`to be protected against BC breaks .

By the way: From checking the code out I've definitely seen much worse coding being released as production-quality code ;)

spekulatius commented:
Hello @mvdbos and @solverat,

I've went through and updated/fixed it as good as possible for now. Here high-level summary of the changes I've added in the last commit:

- First of all: Allow event-distributor v5 and exclude earlier versions. While testing the code, I had to allow upgrading [`symfony/finder`](https://github.com/symfony/finder) to v5. I've not found any breaking changes from this, but it might be advisable to double check this. See https://github.com/mvdbos/php-spider/pull/67/files#diff-b5d0ee8c97c7abd7e3fa29b9a27d1780L23.

- I switched the parameters for the dispatcher in multiple positions in the code as well as the tests. See https://github.com/mvdbos/php-spider/pull/67/files for details please.

- I fixed the `use`-statement to match the actual class location for the `Event`-class in a few cases (examples + `Downloader` and `InMemoryQueueManager`). It's part of the `Contracts`, not `Components` as far as I could see. Please let me know if this isn't correct. See for the changed line: https://github.com/mvdbos/php-spider/pull/67/files#diff-421bb39a0e9a5f5322b5665ac248d5bfL4

- I've added some minor code style to address things I've noted file running the tests/checking the travis report.

Please note: As far as I can see, the currently failing travis check come from an authentication issue with GitHub. See https://travis-ci.org/github/mvdbos/php-spider for more details.

Please let me know what you think of the changes suggest.

Peter

mvdbos commented:
> Thanks for replying to the PR. I agree - it's a major change. From a user' view the interface doesn't change, does it?

True, but it will for example break support for Symfony 3, and probably older versions of Laravel too. This is why we need to be very careful and indeed introduce a new major version for this, so people stuck on those versions don't get breaking changes pushed to them.

mvdbos commented:
> Please note: As far as I can see, the currently failing travis check come from an authentication issue with GitHub.

I tried to fix it yesterday. No luck so far. Don't have much time today, but will try to have another look.

spekulatius commented:
> > Please note: As far as I can see, the currently failing travis check come from an authentication issue with GitHub.
> 
> I tried to fix it yesterday. No luck so far. Don't have much time today, but will try to have another look.

No problem. Just noticed it and thought I should mention it.

spekulatius commented:
> > Thanks for replying to the PR. I agree - it's a major change. From a user' view the interface doesn't change, does it?
> 
> True, but it will for example break support for Symfony 3, and probably older versions of Laravel too. This is why we need to be very careful and indeed introduce a new major version for this, so people stuck on those versions don't get breaking changes pushed to them.

Okay, that's more than fair enough. I leave the versioning to you of course.

spekulatius commented:
Hello @mvdbos,

I understand if you are busy. Just wanted to check in if there is anything else you need from my end to complete this. Thanks :)

Peter

mvdbos commented:
Apologies for the radio silence. It has been a few hectic weeks. 

I have been looking into it and tried to find a way to keep compatibility with Symfony 3, but this turned out not to be trivial. It requires a hack like PHP-CS-Fixer did:

```php
// Since PHP-CS-FIXER is PHP 5.6 compliant we can't always use Symfony Contracts (currently needs PHP ^7.1.3)
// This conditionnal inheritance will be useless when PHP-CS-FIXER no longer supports PHP versions
// inferior to Symfony/Contracts PHP minimal version
if (class_exists(\Symfony\Contracts\EventDispatcher\Event::class)) {
    class Event extends \Symfony\Contracts\EventDispatcher\Event
    {
    }
} else {
    class Event extends \Symfony\Component\EventDispatcher\Event
    {
    }
}
```
And then replace all uses of `GenericEvent` in the spider codebase with this event. I have not gotten around to that yet. If you have time, please feel free to take stab at it.

mvdbos commented:
Oh wait, looks like we are in luck: if I change the method parameter to `GenericEvent`, which I am using everywhere anyway, it all seems to still work with all versions of Symfony. I need to do some more tests. 

Might seem ugly, but since `Event` is also not an interface, changing the type hint to this slightly more specific class is not a big issue for me.

mvdbos commented:
Well, it is done with the changes I have submitted to @spekulatius's PR. Managed to keep as much BC as possible in there, but support for Symfony/Event-Dispatcher v3 is definitely dropped with this. 

@solverat About versioning: turns out that composer treats every version below 1.0 as a major version: https://getcomposer.org/doc/articles/versions.md#caret-version-range-

This means that as long as users specified a version such as `^0.4.0` (which they are hopefully doing this way), they should be fine and not receive any updates beyond 0.4.x.

spekulatius commented:
Hello @mvdbos,

Thanks for having a look and no problem - totally understandable that other things take priority, especially these days!

I've got no Symfony projects and can only help with Laravel (5-7) testing. As for these, it works :) 

You said you pushed changes to this branch? I haven't seen a commit. Or did I misread this?

Cheers,
Peter

mvdbos commented:
> You said you pushed changes to this branch? I haven't seen a commit. Or did I misread this?

I created a PR to your PR here: https://github.com/SpeksForks/php-spider/pull/1

Please test and merge it, so I can merge this. :-)

spekulatius commented:
Ah, okay. I didn't expect a PR to my PR haha :) The new GitHub notifications didn't help with spotting it...

I've updated the commit messages - and the hashs changed... So I couldn't merge it anymore. I should have seen this coming :/  I decided to clean it up a bit and squash my two commits into one. I've added your proposed changes on top. Could you have a look if this works @mvdbos ?

This is the commit with your changes: https://github.com/mvdbos/php-spider/pull/67/commits/05cf0a8ccda2b3c98d5775734adad0875284659f

mvdbos commented on src/VDB/Spider/Downloader/Downloader.php:
```diff
@@ -86,12 +86,12 @@ public function isDownLoadLimitExceeded()
-     * @param null|Event $event
      */
-    private function dispatch($eventName, Event $event = null)
+    private function dispatch(Event $event, $eventName)
```
In both cases (also in Spider) this should be `private function dispatch(GenericEvent $event, $eventName)`.

spekulatius replied:
Okay, I've updated both cases. Could you verify it?

mvdbos commented:
@spekulatius One last request after you fix above review comment: can you rebase this PR on the latest master? In the meantime I fixed another bug there and released a new version.

spekulatius commented:
> @spekulatius One last request after you fix above review comment: can you rebase this PR on the latest master? In the meantime I fixed another bug there and released a new version.

Done. Have a look over the complete PR please - just want to make sure we aren't missing anything :)

mvdbos commented on src/VDB/Spider/Downloader/Downloader.php:
```diff
@@ -2,10 +2,10 @@
 use Symfony\Component\EventDispatcher\EventDispatcher;
 use Symfony\Component\EventDispatcher\EventDispatcherInterface;
 use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Contracts\EventDispatcher\Event;
```
Is this import still required? If not, please remove it.

mvdbos commented on src/VDB/Spider/Spider.php:
```diff
@@ -2,10 +2,10 @@
 use Symfony\Component\EventDispatcher\EventDispatcher;
 use Symfony\Component\EventDispatcher\EventDispatcherInterface;
 use Symfony\Component\EventDispatcher\GenericEvent;
+use Symfony\Contracts\EventDispatcher\Event;
```
Ditto

spekulatius replied:
Yes, you are right - these shouldn't be needed anymore. I've removed them and update the PR @mvdbos

mvdbos commented on composer.json:
```diff
@@ -18,10 +18,10 @@
+        "symfony/css-selector": "^3.0.0||^4.0.0||^5.0.0",
+        "symfony/dom-crawler": "^3.0.0||^4.0.0||^5.0.0",
+        "symfony/finder": "^3.0.0||^4.0.0||^5.0.0",
+        "symfony/event-dispatcher": "^3.0.0||^4.0.0||^5.0.0",
```
I am sorry, I missed this. That is why it would have been more foolproof if you had been able to merge my PR instead of re-implement it. But it seems we have everything now.

This line should read `"symfony/event-dispatcher": "^4.0.0||^5.0.0",`. Note the removal of version 3.

spekulatius replied:
Yes, you are absolutely right. Merging would have been cleaner. I should have avoided this. I hope we aren't missing anything due to this. I've dropped the v3 support for event-dispatcher.

mvdbos commented:
Ok. Looks good. Thanks for all the work!

spekulatius commented:
Awesome, thank you for your working through the issues and helping to get this merged in @mvdbos! Happy I had a chance to contribute and make it a bit better :)
